lib/Target/LLVM/LLVMToDynamicLib.cpp

```diff
@@ -0,0 +1,303 @@
+  SmallVector<std::string, 8> flags = {
+      "ld.lld-17",
+      "-o " + libraryName,
+  };
+
+  // Avoids including any libc/startup files that initialize the CRT as
+  // we don't use any of that. Our shared libraries must be freestanding.
+  flags.push_back("-nostdlib"); // -nodefaultlibs + -nostartfiles
+
+  // Statically link all dependencies so we don't have any runtime deps.
+  // We cannot have any imports in the module we produce.
+  flags.push_back("-static");
+
+  // Generate a dynamic library (ELF type: ET_DYN), otherwise dlopen()
+  // won't succeed on it. This is not incompatible with -static. The GNU
+  // man page for ld, `man ld`, says the following:
+  //
+  //   -static
+  //       Do not link against shared libraries. [...] This option can be
+  //       used with -shared. Doing so means that a shared library is
+  //       being created but that all of the library's external references
+  //       must be resolved by pulling in entries from static libraries.
+  //
+  // While that much is said in the GNU ld man page, the reality is that
+  // out of ld.bfd, ld.gold and ld.lld, only ld.lld actually implements
+  // that. Meanwhile, ld.bfd interprets -static -shared as just -static,
+  // and ld.gold rejects -static -shared outright as "incompatible".
+  flags.push_back("-shared");
+
+  // Strip debug information (only, no relocations) when not requested.
+  if (removeDebugSymbols) {
+    flags.push_back("--strip-debug");
+  }
+
+  // Link all input objects. Note that we are not linking whole-archive as
+  // we want to allow dropping of unused codegen outputs.
+  for (auto &objectFile : objectFileNames) {
+    flags.push_back(objectFile);
+  }
+
+  auto commandLine = llvm::join(flags, " ");
+  if (llvm::failed(runLinkCommand(commandLine)))
+    return llvm::failure();
+  return llvm::success();
+}
+
+llvm::LogicalResult verifyAllLLVM(mlir::ModuleOp module) {
+  auto llvmDialect = module.getContext()->getOrLoadDialect<LLVM::LLVMDialect>();
+
+  bool isAllLLVM = true;
+
+  module.walk([&](Operation *op) {
+    // Allow the module operation itself to pass (builtin.module)
+    if (llvm::isa<mlir::ModuleOp>(op)) {
+      return; // Skip the check for the module operation
+    }
+    // check other operations to make sure they're llvm
+    if (op->getDialect() != llvmDialect) {
+      isAllLLVM = false;
+      llvm::errs() << "Non-LLVM operation found: " << op->getName()
+                   << " at location " << op->getLoc() << "\n";
+    }
+  });
+
+  if (isAllLLVM) {
+    llvm::outs() << "All operations belong to the LLVM dialect.\n";
+    return llvm::success();
+  } else {
+    llvm::errs() << "Module contains non-LLVM dialect operations.\n";
+    return llvm::failure();
+  }
+}
+
+std::optional<std::vector<uint8_t>>
+compileAndLinkToSharedLibrary(llvm::Module &module,
+                              llvm::LLVMContext &context) {
+  const auto tmpDirName = createTempDir();
```

Comment:
@nsmithtt or others: a question I had is whether we should have some kind of flag to say "don't delete temporary .o /.so files we generate" for debugging, or what?  Probably would usually want to clean them up, but in some cases they might be useful debugging tools I guess.  I'm not familiar with how we think about debug flags in general in our code

Reply:
Yeah, I would think either that or making it a pipeline argument that defaults to "delete" would be the 2 options that make sense to me, no strong preference one way vs the other myself

Reply:
Yeah I like pass pipeline option even better